xapian-core/matcher/synonympostlist.cc

```diff
@@ -90,6 +90,8 @@ SynonymPostList::get_weight(Xapian::termcount doclen,
 	    }
 	    if (wdf > doclen) wdf = doclen;
 	}
+	// Set wdfdocmax to doclen in case of a synonym.
+	wdfdocmax = doclen;
```

Comment:
This dosn't seem quite right.

One problem is it only currently gets executed if `want_wdf` is set, but there's no inherent reason why a weighting scheme which uses wdfdocmax would also use wdf so this should be outside that `if`.

Another problem is that `doclen` may not have been read here (a weighting scheme could use wdfdocmax, but not doclen, and in that case the doclen won't be set unless the `(!wdf_disjoint && wdf > doclen_lower_bound)` condition above was true.  The approach we currently take to this for clamping wdf to doclen can be seen just above, though unlike that situation, for wdfdocmax we'll always want doclen available here so it's probably better to have `WDFDOCMAX` automatically enable `DOCLEN` when weighting a synonym (which is handled by a separate overloaded form of `Weight::init_()`).

I've pushed a commit to your branch which adds a testcase to demonstrate these problems - it fails currently, but I think it should pass for all 3 flag combinations tested.